src/tactic/core.lean

```diff
@@ -692,6 +694,45 @@ meta def lock_tactic_state {α} (t : tactic α) : tactic α
        | result.exception msg pos s' := result.exception msg pos s
 end
 
+meta def type_cast (α : Type u) [reflected α] (n : name) : tactic α :=
+eval_expr α (expr.const n [])
+
+meta def attempt_refl (lhs rhs : expr) : tactic expr :=
```

Comment:
Would this be better done with `solve_aux`? I'm not sure that the subgoal handling here is the same though.
```lean
meta def attempt_refl (lhs rhs : expr) : tactic expr :=
lock_tactic_state $ 
do tgt ← to_expr ``(%%lhs = %%rhs),
   prod.snd <$> solve_aux tgt `[refl, done]
```

Also, this could use a doc string.

Reply:
What does `lock_tactic_state` do? Also, I think we should start advocating for using `tactic.mk_app` instead of `to_expr` whenever possible. Here, ``mk_app `eq [lhs,rhs]`` does nicely.

Reply:
`lock_tactic_state` should get the result of the given tactic without modifying the tactic state. Between this and the `apply_core` config, I'm not sure exactly what Keeley is trying to do with subgoals/metavars here, which is why my suggestion might not be a drop-in replacement.

I agree about using `mk_app`.

Reply:
I see. We don't need it when using `solve_aux` then, do we?

Reply:
It can make a difference in the presence of metavars. I'm not sure if that's a concern here. Compare:
```lean
run_cmd
do m ← mk_meta_var `(ℕ),
   e ← mk_app `eq [m, `(0)],
   (a, b) ← tactic.lock_tactic_state $ solve_aux e `[refl],
   m ← instantiate_mvars m,
   trace m.to_raw_fmt

run_cmd 
do m ← mk_meta_var `(ℕ),
   e ← mk_app `eq [m, `(0)],
   (a, b) ← solve_aux e `[refl],
   m ← instantiate_mvars m,
   trace m.to_raw_fmt
```

Reply:
What you seem to be pointing out here is that `lock_tactic_state` allows you roll back a meta variable assignment. When the goal is a meta variable allocated inside of the `lock_tactic_state` call, it can also roll back the allocation of meta variable. That sounds like inviting trouble

Reply:
I'm just pointing out that `lock_tactic_state` can make a difference when using `solve_aux`. If I were writing `attempt_refl`, I would do it with `solve_aux` and without `lock_tactic_state`. But the various rewrite_search tactics seem to be careful about modifying metavar assignments and I don't know exactly what Keeley's use case is here.

Reply:
@robertylewis Just reading this for the first time. :) `attempt_refl` is just meant to try to prove `lhs = rhs` by an applications of `eq.refl`. My understanding of the implementation of `tactic.reflexivity` is that your code snippet is more elegant and functionally equivalent to what I had written, notwithstanding me wanting to instantiate metavars after running `solve_aux`.

Reply:
On the other hand, I do not understand the other point which you are making. My understanding of your first `run_cmd` is that it cannot possibly do anything, since the bound value of `m` is just thrown away. Instead I am using `instantiate_mvars` inside the `lock_tactic_state` block to extract the `expr`-value bound to the metavar.

Reply:
Well, when I say "can't do anything", I mean that the last two lines of the block will be unaffected by running any tactic at all under the `lock_tactic_state`.

Reply:
I think the only point I was trying to make with the `run_cmd` blocks was that running `solve_aux` without using `lock_tactic_state` can modify the metavariable assignments. It's not really relevant to the PR.